dom.bs

```diff
@@ -2652,6 +2652,50 @@ of a <var>node</var> into a <var>parent</var> before a <var>child</var>, run the
  <!-- Technically this is post-insert. -->
 </ol>
 
+<p>To <dfn export for=Node id=concept-node-ensure-pre-move-validity>ensure pre-move validity</dfn>
+of a <var>node</var> into a <var>parent</var> before a <var>child</var>, run these steps:
+
+<ol>
+ <li><p>If either <var>parent</var> or <var>node</var> are not <a>connected</a>, then
+ <a>throw</a> a "{{HierarchyRequestError!!exception}}" {{DOMException}}.</p></li>
```

Comment:
Hmm, why do we have this limitation. Why not allow move within the same tree? Basically why the condition about shadow-including root isn't enough?

Reply:
I believe we discussed this at TPAC. The idea is that `moveBefore()` is used to preserve state, so in cases where state cannot be preserved, we should not just fall back to usual `insertBefore()`-style behavior. Instead, we'll throw and let the developer handle this case themselves. Basically we let `moveBefore()` succeed only when we're sure we can perform a state-preserving move, and when disconnected destinations or origins are involved, the state cannot be preserved.

Reply:
I would definitely expect a move operation to succeed if one stays within a same tree, whether or not your connected to document.

Reply:
I don't see how connected <> disconnected would work. It would need a completely different code path, no?

I can see connected <> connected and disconnected <> disconnected. Someone (either Noam or Dominic) argued against disconnected <> disconnected earlier, but it seems reasonable to allow as well.

Reply:
I'm ok with both. One can say that disconnected<>disconnected doesn't have any state to preserve, so whether we throw or do the same as insertBefore is a judgement call.

Reply:
We should not do the same as insertBefore. It should run the moving steps (whether anyone cares about those for disconnected is a different question).

Reply:
Sure, I mean that it should be observably the same in the disconnected<>disconnected case.

Reply:
I've made this PR support both the `connected->connected` and `disconnected->disconnected` case, as of https://github.com/whatwg/dom/pull/1307/commits/2bbc8346167310cc3e7f8d2ea65e5151c9cc219f. I think that's sufficient, and we've heard that developers want this too, so I think that's enough to resolve things here.

Reply:
Disconnected -> connected could theoretically also been seen as stated preserving since the disconnected node didn't have any previous state so there is nothing to preserve.

Connected -> disconnected moves could destroy the state though so should definitely not be allowed.

Reply:
> Disconnected -> connected could theoretically also been seen as stated preserving since the disconnected node didn't have any previous state so there is nothing to preserve.

I think I disagree. The state could be null `contentDocument`, lack of focus, no form owner, and more etc. That "no state" is still a kind of state, and it'd either be:
 - **Changed** by becoming connected into a document, and running the many hooks that the platform does during that milestone — some of which initialize iframe setup, run script, etc. In this case, move just becomes insertion, which we don't want to do because the two are distinct primitives.
 - **Not changed**, by just inserting the element into the DOM and leaving it kind of dead/void. This would be completely new and I think kind of weird. For example `parent.moveBefore(disconnectedIframe, null)` would make `disconnectedIframe` finally connected, but with a null content document because we never initialized the setup. In that case, it feels like there was no reason to *move* it into the DOM in the first place. Same goes for other things with form owners, executable `<script>`s, etc. I'm not sold that this is a real use case.